### PR TITLE
subplot titles respecting row_width and column_width

### PR DESCRIPTION
**Pull request by Kully**
issue: https://github.com/plotly/plotly.py/issues/1229

Going to write some tests that ensure that subtitles are in the right place w/ row_/column_width, if none exist already.

Still fixing broken tests

newplot

newplot 1

### CODE REVIEW

File: plotly/tests/test_core/test_tools/test_make_subplots.py [hunk 1]
```diff
@@ -2155,3 +2156,57 @@ def test_large_columns_no_errors(self):
         fig = tls.make_subplots(100, 1,
                                 vertical_spacing=v_space,
                                 specs=[[{'is_3d': True}] for _ in range(100)])
+
+    def test_row_width_and_column_width(self):
+
+        expected = Figure({
```
[the remark this answers is not in the document]
Comment (Kully) [reply plotly/tests/test_core/test_tools/test_make_subplots.py]: just a png in the file? I am not aware of image tests for this repo but I could be wrong

File: plotly/tools.py [hunk 2]
```diff
@@ -1080,6 +1079,7 @@ def _get_anchors(r, c, x_cnt, y_cnt, shared_xaxes, shared_yaxes):
         return x_anchor, y_anchor
 
     list_of_domains = []  # added for subplot titles
+    list_of_domains_complete = []  # hold onto every domain used, even if shared axes
```
[the remark this answers is not in the document]
Comment (Kully) [reply plotly/tools.py]: nope - I'll ⚡️

File: plotly/tools.py [hunk 3]
```diff
@@ -1323,20 +1323,33 @@ def _pad(s, cell_len=cell_len):
...
             subtitle_pos_y.append(y_domains[1])
+
     # If shared_axes is True the domin of each subplot is not returned so the
     # title position must be calculated for each subplot
     else:
-        subtitle_pos_x = [None] * cols
-        subtitle_pos_y = [None] * rows
-        delt_x = (x_e - x_s)
+        x_dom = [layout[k]['domain'] for k in sorted(layout.to_plotly_json().keys()) if 'xaxis' in k]
+        y_dom = [layout[k]['domain'] for k in sorted(layout.to_plotly_json().keys()) if 'yaxis' in k]
```
[the remark this answers is not in the document]
Comment (Kully) [reply plotly/tools.py]: great catch - I'll make sure it's full proof

File: plotly/tests/test_core/test_tools/test_make_subplots.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jonmmease) [reply plotly/tests/test_core/test_tools/test_make_subplots.py]: No, I just meant in the GitHub PR so I could see what it looks like 🙂

[the remark this answers is not in the document]
Comment (Kully) [reply plotly/tests/test_core/test_tools/test_make_subplots.py]: newplot

File: plotly/tools.py [hunk 4]
```diff
@@ -1323,20 +1323,33 @@ def _pad(s, cell_len=cell_len):
...
-        for index in range(rows):
-            subtitle_pos_y[index] = (1 - ((y_e + vertical_spacing) * index))
-        subtitle_pos_y *= cols
-        subtitle_pos_y = sorted(subtitle_pos_y, reverse=True)
+            subtitle_pos_x = []
+            for x_domains in x_dom:
+                subtitle_pos_x.append(sum(x_domains) / 2)
+            subtitle_pos_x *= rows
+
+        if shared_yaxes:
```
[the remark this answers is not in the document]
Comment (Kully) [reply plotly/tools.py]: no, i'll write one

**Comment by jonmmease** [comment]
Code looks good.  Last thing, could you add a screenshot (in the PR) and test for this example from the documentation: https://plot.ly/python/subplots/#custom-sized-subplot-with-subplot-titles

I want to make sure this doesn't break cases with `colspan` and then 🔒 it in.  Thanks!

**Comment by Kully** [comment]
> could you add a screenshot (in the PR) and test for this example from the documentation

Looks good to me.

newplot 1

**Comment by jonmmease** [comment]
Great! Just make it a test and we're good to go.  I'll merge it in tonight. Thanks!

**Comment by Kully** [comment]
> Just make it a test and we're good to go

we already have a test that already tests this example except for the data: https://github.com/plotly/plotly.py/blob/master/plotly/tests/test_core/test_tools/test_make_subplots.py#L2086-L2090

Do you want something that also checks the data or is't it good enough that the layout is being tested?

**Comment by jonmmease** [comment]
No, that covers it. Didn't realize we had this one already! 💃

**Pull request merged**